BJS2-66310-delete-expired-ad-posts

gulnaz21 opened the pull request.

gulnaz21 committed:
BJS2-66310-delete-expired-ad-posts без тестов

gulnaz21 committed:
добавлены тесты

gulnaz21 committed:
метод для удаления рекламы с аннотацией @Async вынесен в отдельный класс

gulnaz21 committed:
добавлен JavaDoc

marsel-mkh commented on src/main/java/faang/school/postservice/config/AdCleanupExecutorConfig.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.concurrent.Executor;
+
+@Configuration
```
Классное решение вынести в конфиг

marsel-mkh commented on src/main/java/faang/school/postservice/repository/ad/AdRepository.java:
```diff
@@ -13,4 +13,7 @@ public interface AdRepository extends CrudRepository<Ad, Long> {
 
     List<Ad> findAllByBuyerId(long buyerId);
+
+    @Query("SELECT a FROM Ad a WHERE a.endDate < CURRENT_TIMESTAMP OR a.appearancesLeft = 0")
```
Тут обязателен query? хибернейт вроде может сделать из без него

gulnaz21 replied:
тут же с OR запрос, думаю лучше прописать его вручную

marsel-mkh commented on src/main/java/faang/school/postservice/service/ad/AdCleanupAsyncService.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @param expiredAds список рекламы для удаления
+     */
+    @Async("adCleanupExecutor")
```
Можно добавить логирование и try-catch

marsel-mkh replied:
Хорошо было бы добавить транзакции, но транзакции и асинк плохо вместе работают, так что можно было бы вынести в отдельный класс метод с транзакциями. Если не захочешь сильно менять архитектуру, то можно попробовать использовать TransactionTemplate либо изменять пропоганацию в транзакциях, но по этому поводу лучше почитать

gulnaz21 committed:
Добавлен отдельный класс для удаления рекламы транзакциями

RinatBeybutov commented on src/main/java/faang/school/postservice/model/ad/Ad.java:
```diff
@@ -11,14 +11,18 @@
 @Data
 @Builder
 @Entity
+@AllArgsConstructor
```
давай на этот класс добавим java doc, что-то вроде "Сущность рекламы"

RinatBeybutov commented on src/main/java/faang/school/postservice/service/ad/AdCleanupAsyncService.java:
```diff
@@ -0,0 +1,36 @@
+            adTransactionalService.deleteAdsBatchInTransaction(expiredAds);
+        } catch (Exception e) {
+            log.error("Failed to cleanup {} ads: {}", expiredAds.size(), e.getMessage());
+            throw new RuntimeException("Async cleanup failed for batch", e);
```
а кто эту ошибку будет обрабатывать?

RinatBeybutov commented on src/main/java/faang/school/postservice/service/ad/AdTransactionalService.java:
```diff
@@ -0,0 +1,23 @@
+    private final AdRepository adRepository;
+
+    @Transactional
+    public void deleteAdsBatchInTransaction(List<Ad> ads) {
```
![image](https://github.com/user-attachments/assets/69836c84-6725-409a-908b-58ad313479d2)

По дефолту у транзакции propagation стоит в значении required - если есть транзакция, то новая не создается. В AdCleanupService уже создалась транзакция, а здесь не создается. В такой конфигурации создание класса AdTransactionalService не особо имеет смысл. Тогда можно в AdTransactionalService просто метод написать. 

В чем была задумка такой конфигурации классов?

gulnaz21 committed:
fix:
- Убрано выбрасывание исключения в cleanupExpiredAdsAsync
- Удалена избыточная @Transactional
- Добавлен JavaDoc для класса Ad

gulnaz21 committed:
fix:
- исправлен тест для AdCleanupAsyncService

RinatBeybutov commented:
задачу закрыл, нужно запушить обновления и смержить

gulnaz21 committed:
Merge branch 'murloc-master-stream9' into feature/BJS2-66310-delete-expired-ad-posts

# Conflicts:
#	src/main/resources/application.yaml
